src/bp/core/services/middleware/state-manager.ts

```diff
@@ -93,14 +101,62 @@ export class StateManager {
     state.bot = await this.kvs.forBot(event.botId).get(this.BOT_GLOBAL_KEY)
     state.__stacktrace = []
 
+    this.boxWorkflowVariables(state.session.workflows)
+
     Object.defineProperty(state, 'workflow', {
       get() {
         return state.session.workflows[state.session.currentWorkflow!]
       }
     })
+
+    // This can be used to set a variable on the current workflow, or on a specific workflow
+    state.setVariable = (
+      name: string,
+      value: any,
+      type: string,
+      options?: { nbOfTurns: number; specificWorkflow?: string }
+    ) => {
+      const wf = options?.specificWorkflow ? state.session.workflows[options?.specificWorkflow] : state.workflow
+      if (!wf) {
+        return
+      }
+
+      const BoxedVar = this._variables?.find(x => x.type === type)
+      if (BoxedVar) {
+        wf.variables[name] = new BoxedVar({
+          nbOfTurns: options?.nbOfTurns ?? 10,
+          value
+        })
```

Comment:
That's pretty fancy I didn't know you could do that with typescript. If I understand correctly you're going through a list of provided types, checking which type has the static property type with the value you want, and then calling the constructor of this type using the type as a variable. That's a lot of reflection!

Reply:
It's standard javascript. Basically, you create your variable definition as a class. If you return the class via an http call (for ex: to the studio), it will only send the static members.

Once you create an instance of that class, you have access to its members. I thought it made sense since everything is contained in the class itself.

Let me know if you have a better solution for that

Reply:
Nope I actually really like this approach 👍